torch/ao/sparsity/_experimental/activation_sparsifier/README.md

```diff
@@ -0,0 +1,75 @@
+3. **Masking**: The reduced activations are then passed into a user-defined ```mask_fn``` to compute the mask.
+
+Essentially, the high level idea of computing the mask is
+
+```
+aggregated_tensor = aggregate_fn([activation for activation in all_activations])
+mask = mask_fn(reduce_fn(aggregated_tensor))
+```
+
+*The activation sparsifier also supports per-feature/channel sparsity. This means that a desired set of features in an activation can be also pruned. The mask will be stored per feature.*
+
+```
+for i in range(len(features)):
+    aggregated_tensor_feature = aggregate_fn([activation[features[i]] for activation in all_activations])
+    mask[i] = mask_fn(reduce_fn(aggregated_tensor_feature))
+```
+
+## Implementation Details
+The activation sparsifier attaches itself to a set of layers in a model and then attempts to sparsify the activations flowing through them. *Attach* means registering a [```forward_pre_hook()```](https://pytorch.org/docs/stable/_modules/torch/nn/modules/module.html#register_forward_pre_hook) to the layer.
+Let's go over the 3 steps again -
+1. **Aggregation**: The activation of aggregation happens by attaching a hook to the layer that specifically applies and stores the aggregated data. The aggregation happens per feature, if the features are specified, otherwise it happens on the entire tensor.
+2. **Reduce**: This is initiated once the ```step()``` is called. The ```reduce_fn()``` is called on the aggregated tensor.
+3. **Masking**: The computation of the mask happens immediately after the reduce operation. The ```mask_fn()``` is applied on the reduced tensor. Again, this happens per-feature, if the features are specified.
+
+## API Design
+```ActivationSparsifier```: Attaches itself to a model layer and sparsifies the activation flowing through that layer. The user can pass in the default ```aggregate_fn```, ```reduce_fn``` and ```mask_fn```. Additionaly, ```features``` and ```feature_dim``` are also accepted.
+
+```register_layer```: Registers a layer for sparsification. Specifically, registers ```forward_pre_hook()``` that performs aggregation.
+
+```step```: For each registered layer, applies the ```reduce_fn``` on aggregated activations and then applies ```mask_fn``` after reduce operation.
+
+```squash_mask```: Unregisters aggregate hook that was applied earlier and registers sparsification hooks if ```attach_sparsify_hook=True```. Sparsification hooks applies the computed mask to the activations before it flows into the registered layer.
+
+## Example
+```
+# Fetch model
+model = SomeModel()
+
+# define some aggregate, reduce and mask functions
+def aggregate_fn(tensor1, tensor2):
+    return tensor1 + tensor2
+
+def reduce_fn(tensor):
+    return tensor.mean(dim=0)
+
+def mask_fn(data):
+    return torch.eye(data.shape).to(data.device)
+
+# define activation sparsifier
+act_sparsifier = ActivationSparsifier(model, aggregate_fn, reduce_fn, mask_fn)
```

Comment:
nit: I am not sure if PyTorch supports enforcing the "keyword args only", but I think we should at least *encourage* passing the functions using named args.